File: OmniScript/cs/OmniScript/HostPlatform.cs
// =============================================================================
// <copyright file="Adapter.cs" company="LiveAction, Inc.">
//  Copyright (c) 2018-2021 Savvius, Inc. All rights reserved.
//  Copyright (c) 2015-2018 Savvius, Inc. All rights reserved.
// </copyright>
// =============================================================================

namespace Savvius.Omni.OmniScript
{
    using System;
    using System.Collections.Generic;

    public class HostPlatform
    {
        private const uint CaptureMagic = 0xABCD;
        private const uint VigilMagic = 0x5653;
        private const uint InSightMagic = 0x4953;

        public static readonly String DefaultAuth = "Default";
        public static readonly String WindowsAuth = "Default";
        public static readonly String LinuxAuth = "ThirdPartyAuth";
        public static readonly String MacAuth = "ThirdPartyAuth";

        public enum EngineKind { Unknown, Capture, Vigil, InSight };
        public enum OperatingSystem { Unknown, Windows, Linux, Mac };

        private readonly Dictionary<uint, EngineKind> KindMap = new Dictionary<uint, EngineKind>()
        {
            { CaptureMagic, EngineKind.Capture },
            { VigilMagic, EngineKind.Vigil },
            { InSightMagic, EngineKind.InSight }
        };

        public EngineKind Kind { get; set; }
        public uint Magic { get; set; }
        public uint Major { get; set; }
        public uint Minor { get; set; }
        public OperatingSystem OS { get; set; }

        public HostPlatform(uint magic, uint version, byte os)
        {
            this.Magic = magic;
            this.Kind = (KindMap.ContainsKey(magic)) ? KindMap[magic] : EngineKind.Unknown;
            this.Major = version & 0x00FF;   // 0x0501 == 1.5
            this.Minor = (version & 0x0FF00) >> 8;
            
            switch (os)
            {
                case 3:
                    this.OS = OperatingSystem.Windows;
                    break;

                case 70:
                    this.OS = OperatingSystem.Mac;
                    break;

                case 100:
                    this.OS = OperatingSystem.Linux;
                    break;

                default:
                    this.OS = OperatingSystem.Unknown;
                    break;
            }
        }

        static public String GetAuth(HostPlatform platform, String auth)
        {
            if (String.IsNullOrEmpty(auth))
            {
                if (platform != null) {
                    switch (platform.OS)
                    {
                        case OperatingSystem.Windows:
                            return HostPlatform.WindowsAuth;

                        case OperatingSystem.Linux:
                            return HostPlatform.LinuxAuth;

                        case OperatingSystem.Mac:
                            return HostPlatform.MacAuth;

                        default:
                            return HostPlatform.DefaultAuth;
                    }
                }
                return HostPlatform.DefaultAuth;
            }

            String lowauth = auth.ToLower().Replace(' ', '-');
            if (lowauth == "default")
            {
                return HostPlatform.WindowsAuth;
            }
            else if ((lowauth == "thirdparty") || (lowauth == "third-party"))
            {
                return HostPlatform.LinuxAuth;
            }

            return HostPlatform.DefaultAuth;
        }

        public ulong GetProtocol()
        {
            return (this.Magic << 16) | ((this.Minor << 8) & 0x0FF) | (this.Major & 0x0FF);
        }
    }
}

File: OmniScript/py/OmniScript/omniscript/hostplatform.py
"""HostPlatform class.
"""
#Copyright (c) WildPackets, Inc. 2015. All rights reserved.

import os

import omniscript

CAPTURE_MAGIC = 0XABCD
VIGIL_MAGIC = 0X5653
INSIGHT_MAGIC = 0X4953

class HostPlatform(object):
    """Information about the engine's operating system.
    """

    kind = 0
    """The kind of engine. One of the Engine Kind constants."""

    major = 0
    """The major part of the version. The 1 in 1.2."""

    minor = 0
    """The minor part of the version. The 2 in 1.2."""

    os = 0
    """The kind of operating system. One of the
    Operating System constants
    """
    
    auth_strings = ["Default", "ThirdPartyAuth"]
    """List of auth strings. One of the 
    Operating System constants.
    """

    _magic = 0
    """The original magic value."""

    _kind_map = {
        CAPTURE_MAGIC : omniscript.ENGINE_KIND_CAPTURE,
        VIGIL_MAGIC : omniscript.ENGINE_KIND_VIGIL,
        INSIGHT_MAGIC : omniscript.ENGINE_KIND_INSIGHT
        }

    def __init__(self, magic, version, os):
        self._magic = magic
        self.kind = HostPlatform._kind_map[magic]
        # 0x0501 is 1.5
        self.major = version & 0x0FF
        self.minor = (version & 0x0FF00) >> 8
        if os == 3:
            self.os = omniscript.OPERATING_SYSTEM_WINDOWS
        elif os == 100:
            self.os = omniscript.OPERATING_SYSTEM_LINUX
        elif os == 70:
            self.os = omniscript.OPERATING_SYSTEM_MAC
        else:
            self.os = omniscript.OPERATING_SYSTEM_UNKNOWN

    def version(self):
        return '%d.%d' % (self.major, self.minor)

    def get_protocol(self):
        return (self._magic << 16) | (self.minor) << 8 | self.major

    @staticmethod
    def get_auth(platform, auth):
        if not auth:
            if platform:
                if platform.os == omniscript.OPERATING_SYSTEM_WINDOWS:
                    return HostPlatform.auth_strings[0]
                if platform.os == omniscript.OPERATING_SYSTEM_LINUX:
                     return HostPlatform.auth_strings[1]
            return HostPlatform.auth_strings[0]

        lowauth = auth.lower().replace(' ', '-')
        if lowauth == "default":
            return HostPlatform.auth_strings[0]
        if lowauth == "thirdparty" or lowauth == "third-party":
            return HostPlatform.auth_strings[1]
        return HostPlatform.auth_strings[0]
